File: apps/ios/AgentCompanion/Models/Alert.swift
import Foundation

/// A security detector alert linked to an event.
struct SecurityAlert: Codable, Identifiable, Equatable {
    let id: UUID
    let eventID: UUID
    let detectorName: String
    let severity: Severity
    let explanation: String
    let recommendedAction: String
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case eventID = "event_id"
        case detectorName = "detector_name"
        case severity, explanation
        case recommendedAction = "recommended_action"
        case createdAt = "created_at"
    }
}

File: backend/app/models/alert.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.event import Severity


class Alert(Base):
    __tablename__ = "alerts"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    event_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("events.id"), nullable=False, index=True
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True
    )
    instance_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("instances.id"), nullable=False, index=True
    )
    detector_name: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    skill_name: Mapped[str] = mapped_column(String(255), nullable=False)
    severity: Mapped[Severity] = mapped_column(Enum(Severity), nullable=False)
    explanation: Mapped[str] = mapped_column(Text, nullable=False)
    recommended_action: Mapped[str] = mapped_column(String(255), nullable=False)
    evidence: Mapped[dict | None] = mapped_column(JSONB)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), index=True)

    event: Mapped["Event"] = relationship(back_populates="alerts")
